File: staff/tools/codegen/templates/common/EnumSerialization.cpp
#cginclude "StringTypes"

std::string SerializeEnum_$(Enum.NsName.!mangle)_ToString(const $(Enum.NsName) eEnumValue)
{
#ifeq($(Enum.Options.*baseType),$($sStringTypes))
  switch (eEnumValue)
  {
#foreach $(Enum.Members)
  case $(Enum.Namespace)$(Member.Name):
    return "$(Member.Value||Member.Name)";
#end
  default:
    STAFF_THROW_ASSERT("Value out of range while serializing enum [$(Enum.NsName)]: " +
                       staff::ToString(eEnumValue));
  };
#else
  return ToString(static_cast<int>(eEnumValue));
#ifeqend
}

DataObject& operator<<(DataObject& rdoParam, const $(Enum.NsName) eEnumValue)
{
  rdoParam.SetText(SerializeEnum_$(Enum.NsName.!mangle)_ToString(eEnumValue));
  return rdoParam;
}

$(Enum.NsName) DeserializeEnum_$(Enum.NsName.!mangle)_FromString(const std::string& sParam)
{
#ifeq($(Enum.Options.*baseType),$($sStringTypes))
#foreach $(Enum.Members)
  if (sParam == "$(Member.Value||Member.Name)")
  {
    return $(Enum.Namespace)$(Member.Name);
  }
  else
#end
  {
    STAFF_THROW_ASSERT("Value out of range while deserializing enum [$(Enum.NsName)]: [" + sParam + "]");
  }
#else
  int nValue = 0;
  FromString(sParam, nValue);
  return static_cast< $(Enum.NsName) >(nValue);
#ifeqend
}

const DataObject& operator>>(const DataObject& rdoParam, $(Enum.NsName)& reEnumValue)
{
  reEnumValue = DeserializeEnum_$(Enum.NsName.!mangle)_FromString(rdoParam.GetText());
  return rdoParam;
}

File: staff/tools/codegen/templates/jsclient/EnumSerialization.js
#ifeq($(Enum.Extern),false) // do not serialize extern type
#cginclude <common/StringTypes>

// enum $(Enum.NsName.!mangle)
var o$(Enum.NsName.!mangle) = staff.object("$(Enum.NsName.!dot)");
#var EnumValue 0
#foreach $(Enum.Members)
o$(Enum.NsName.!mangle).$(Member.Name) =\
#ifneq($(Member.Value),)
 $(Member.Value);
#var EnumValue $(Member.Value.!inc)
#else
 $($EnumValue);
#var EnumValue $($EnumValue.!inc)
#ifeqend
#end

o$(Enum.NsName.!mangle).serialize = function(oElement, oType)
{
#ifeq($(Enum.Options.*baseType),$($sStringTypes))
  var sResult = "<INVALID>";
  switch (oType)
  {
#foreach $(Enum.Members)
  case $(Enum.NsName.!mangle).$(Member.Name):
    sResult = "$(Member.Value||Member.Name)";
    break;
#end
  };
  oElement.setText(sResult)
#else
  oElement.setText(oType)
#ifeqend
  return oElement;
}

o$(Enum.NsName.!mangle).deserialize = function(oElement)
{
#ifeq($(Enum.Options.*baseType),$($sStringTypes))
  var oResult = -1;
  switch (staff.getElementValue(oElement))
  {
#foreach $(Enum.Members)
  case "$(Member.Value||Member.Name)":
    oResult = $(Enum.NsName.!mangle).$(Member.Name);
    break;
#end
  };
  return oResult;
#else
  return parseInt(staff.getElementValue(oElement));
#ifeqend
}

#ifeqend //ifeq((Enum.Extern),false) // do not serialize extern type
